gap/semigroups/semifp.gi

```diff
@@ -636,3 +636,163 @@ function(M)
+                rbracket := -1;
+                # tracks how 'deep' the position of i is in terms of nested
+                # brackets
+                nestcount := 0;
+                i := i + 1;
+                while i <= Size(word) do
+                    if word[i] = '(' then
+                        nestcount := nestcount + 1;
+                    elif word[i] = ')' then
+                        if nestcount = 0 then
+                            rbracket := i;
+                            break;
+                        else
+                            nestcount := nestcount - 1;
+                        fi;
+                    fi;
+                    i := i + 1;
+                od;
+                # if rbracket is not followed by ^ then the value inside the
+                # bracket is appended (recursion is used to remove any brackets
+                # in this value)
+                if rbracket = Size(word) or (not word[rbracket + 1] = '^') then
+                    if product = "" then
+                      product := RemoveBrackets(
+                                 word{[lbracket + 1 .. rbracket - 1]});
+                    else
+                      product := product *
+                                 RemoveBrackets(
+                                 word{[lbracket + 1 .. rbracket - 1]});
+                    fi;
+                # if rbracket is followed by ^ then the value inside the
+                # bracket is appended the given number of time
+                else
+                    i := i + 2;
+                    while i <= Size(word) do
+                        if word[i] in "0123456789" then
+                            i := i + 1;
+                        else
+                            break;
+                        fi;
+                    od;
+
+                    p := Int(word{[rbracket + 2 .. i - 1]});
+                    if p = 0 then
+                      ErrorNoReturn(Concatenation("expected ^ to be followed",
+                      " by a positive integer but found 0"));
+                    fi;
+                    if product = "" then
+                       product := RemoveBrackets(word{[lbracket + 1 ..
+                                                       rbracket - 1]}) ^ p;
+                    else
+                       product := product *
+                                  RemoveBrackets(word{[lbracket + 1 ..
+                                                       rbracket - 1]}) ^ p;
+                    fi;
+                    i := i - 1;
+                fi;
+            fi;
+            i := i + 1;
+        od;
+        return product;
+    end;
+
+    ParseRelation := x -> List(SplitString(x, "="), RemoveBrackets);
+    output := List(SplitString(newinputstring, ","), ParseRelation);
+    if ForAny(output, x -> Size(x) = 1) then
+      ErrorNoReturn(Concatenation("expected a product of free",
+                          " generators but found an empty string"));
```

Comment:
Something more like "expected the second argument to be a string representing a list of relations, but found an empty relation" or whatever the actual correct thing is?